File: branches/MPLRServer/DataParsers/Guild.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace MPLRServer
{
    public class Guild
    {
        public class GuildSkill
        {
            public int SkillID { get; set; }
            public short Level { get; set; }
            public long BoughtAt { get; set; }
            public string BoughtBy { get; set; }
            public string Unknown { get; set; }
        }

        public class GuildMember
        {
            public int CharacterID { get; set; }
            public byte Rank { get; set; }
            public int Contribution { get; set; }
        }

        public int ID { get; private set; }
        public int Points { get; private set; }
        public int AllianceID { get; private set; }
        public byte Capacity { get; private set; }
        public string Name { get; private set; }
        public string Notice { get; private set; }
        public string[] Ranks { get; private set; }

        public byte Players { get; private set; }

        public short Background { get; private set; }
        public byte BackgroundColor { get; private set; }
        public short Foreground { get; private set; }
        public byte ForegroundColor { get; private set; }

        public List<GuildSkill> Skills { get; private set; }

        public List<GuildMember> Members { get; private set; }

        public Guild()
        {
            Ranks = new string[5];
        }

        public void Decode(MaplePacket pPacket)
        {
            ID = pPacket.ReadInt();
            Name = pPacket.ReadString();

            for (byte i = 0; i < 5; i++)
                Ranks[i] = pPacket.ReadString();

            Players = pPacket.ReadByte();

            Members = new List<GuildMember>();
            for (byte i = 0; i < Players; i++)
            {
                var id = pPacket.ReadInt(); // Player ID

                Members.Add(new GuildMember() { CharacterID = id });
            }

            for (byte i = 0; i < Players; i++)
            {
                GuildMember tmp = Members[i];

                pPacket.ReadString(13); // Name
                pPacket.ReadInt(); // Job ID
                pPacket.ReadInt(); // Level OR -1
                tmp.Rank = (byte)pPacket.ReadInt(); // Guild Rank
                pPacket.ReadInt(); // Online
                pPacket.ReadInt(); // Alliance Rank
                tmp.Contribution = pPacket.ReadInt(); // Contribution
            }

            Capacity = (byte)pPacket.ReadInt();
            Background = pPacket.ReadShort();
            BackgroundColor = pPacket.ReadByte();
            Foreground = pPacket.ReadShort();
            ForegroundColor = pPacket.ReadByte();

            Notice = pPacket.ReadString();

            Points = pPacket.ReadInt();
            pPacket.ReadInt(); // Today's points
            AllianceID = pPacket.ReadInt();

            pPacket.ReadByte(); // Unk
            pPacket.ReadShort(); // Herp

            Skills = new List<GuildSkill>();
            var skills = pPacket.ReadShort(); // Buffs
            for (var i = 0; i < skills; i++)
            {
                GuildSkill gb = new GuildSkill()
                {
                    SkillID = pPacket.ReadInt(),
                    Level = pPacket.ReadShort(),
                    BoughtAt = pPacket.ReadLong(),
                    BoughtBy = pPacket.ReadString(),
                    Unknown = pPacket.ReadString()
                };

                Skills.Add(gb);
            }
        }

        public void Save(byte pWorldID)
        {
            using (InsertQueryBuilder guildTable = new InsertQueryBuilder("guilds"))
            {
                guildTable.OnDuplicateUpdate = true;
                guildTable.AddColumn("id", false);
                guildTable.AddColumn("world_id", true);
                guildTable.AddColumn("name", true);
                guildTable.AddColumn("notice", true);
                guildTable.AddColumn("rank1", true);
                guildTable.AddColumn("rank2", true);
                guildTable.AddColumn("rank3", true);
                guildTable.AddColumn("rank4", true);
                guildTable.AddColumn("rank5", true);
                guildTable.AddColumn("capacity", true);
                guildTable.AddColumn("emblem_bg", true);
                guildTable.AddColumn("emblem_bg_color", true);
                guildTable.AddColumn("emblem_fg", true);
                guildTable.AddColumn("emblem_fg_color", true);
                guildTable.AddColumn("points", true);
                guildTable.AddColumn("alliance_id", true);

                guildTable.AddRow(ID, pWorldID, Name,
                    Notice,
                    Ranks[0],
                    Ranks[1],
                    Ranks[2],
                    Ranks[3] == "" ? null : Ranks[3],
                    Ranks[4] == "" ? null : Ranks[4],
                    Capacity,
                    Background,
                    BackgroundColor,
                    Foreground,
                    ForegroundColor,
                    Points,
                    AllianceID
                    );

                guildTable.RunQuery();
            }

            // Delete members first

            MySQL_Connection.Instance.RunQuery("DELETE FROM guild_members WHERE guild_id = " + ID);

            using (InsertQueryBuilder guildMembersTable = new InsertQueryBuilder("guild_members"))
            {
                guildMembersTable.OnDuplicateUpdate = true;
                guildMembersTable.AddColumn("guild_id", true);
                guildMembersTable.AddColumn("character_id", false); // Switching guild huh?
                guildMembersTable.AddColumn("rank", true);
                guildMembersTable.AddColumn("contribution", true);

                foreach (var member in Members)
                {
                    guildMembersTable.AddRow(
                        ID,
                        member.CharacterID,
                        member.Rank,
                        member.Contribution
                        );
                }

                if (guildMembersTable.RowCount > 0)
                    MySQL_Connection.Instance.RunQuery(guildMembersTable.ToString());
            }

            using (InsertQueryBuilder guildSkillsTable = new InsertQueryBuilder("guild_skills"))
            {
                guildSkillsTable.OnDuplicateUpdate = true;
                guildSkillsTable.AddColumn("guild_id", false);
                guildSkillsTable.AddColumn("skill_id", false);
                guildSkillsTable.AddColumn("level", true);
                guildSkillsTable.AddColumn("bought_by", true);
                guildSkillsTable.AddColumn("bought_at", true);
                guildSkillsTable.AddColumn("unk", true);

                foreach (var skill in Skills)
                {
                    guildSkillsTable.AddRow(
                        ID,
                        skill.SkillID,
                        skill.Level,
                        skill.BoughtBy,
                        skill.BoughtAt,
                        skill.Unknown
                        );
                }

                if (guildSkillsTable.RowCount > 0)
                    MySQL_Connection.Instance.RunQuery(guildSkillsTable.ToString());
            }
        }
    }
}

File: trunk/website/inc/classes/guild.php
<?php
require_once __DIR__.'/../database.php';

class Guild {
	public $name, $id, $world_id, $world_name, $notice, $ranks, 
		$capacity,
		$points, $alliance_id,
		$icon_bg, $icon_bgc, $icon_fg, $icon_fgc;
	public $members;
		
	public function LoadByName($name, $world) {
		global $__database;
		$q = $__database->query("
SELECT
	world_data.world_name,
	guilds.*
FROM
	`guilds`
LEFT JOIN 
	`world_data`
	ON
		world_data.world_id = guilds.world_id
WHERE 
	guilds.name = '".$__database->real_escape_string($name)."'
AND
	world_data.world_name = '".$__database->real_escape_string($world)."'");
		if ($q->num_rows == 0) {
			return false;
		}
		
		$row = $q->fetch_assoc();
		$this->id = $row['id'];
		$this->world_id = $row['world_id'];
		$this->world_name = $row['world_name'];
		$this->name = $row['name'];
		$this->notice = $row['notice'];

		$this->ranks = array();
		for ($i = 1; $i <= 5; $i++)
			$this->ranks[$i - 1] = isset($row['rank'.$i]) ? $row['rank'.$i] : '-';

		$this->icon_bg = $row['emblem_bg'];
		$this->icon_bgc = $row['emblem_bg_color'];
		$this->icon_fg = $row['emblem_fg'];
		$this->icon_fgc = $row['emblem_fg_color'];
		
		$this->points = $row['points'];
		$this->alliance_id = $row['alliance_id'];
		$this->capacity = $row['capacity'];
		
		$q->free();
		
		$this->LoadCharacterList();
		
		return true;
	}
	
	private function LoadCharacterList() {
		global $__database;
		
		$q = $__database->query("
SELECT
	characters.name,
	guild_members.rank,
	guild_members.alliance_rank,
	guild_members.contribution
FROM
	`guild_members`
LEFT JOIN 
	`characters`
	ON
		characters.id = guild_members.character_id
WHERE 
	guild_members.guild_id = ".$this->id."
AND
	guild_members.world_id = ".$this->world_id."
AND
	characters.name IS NOT NULL
ORDER BY
	rank ASC");

		$this->members = array();
		while ($row = $q->fetch_assoc()) {
			$this->members[] = $row;
		}
	}
}


?>